### PR TITLE
Fix handling explicit nil values

### PR DESCRIPTION
**Pull request by ccdredzik**
Should solve #80

### CODE REVIEW

File: spec/rails_integration_spec.rb [hunk 1]
```diff
@@ -106,14 +106,12 @@ def prepare_params(params)
   end
 
   describe "nested_array" do
-    it "responds with a 400 when the nested array is not supplied properly" do
-      params = {
-        'filter' => 'state'
-      }
+    it "responds with a 200 when the nested array is provided as nil" do
+      expect { get :nested_array, **prepare_params({ filter: { state: nil } }) }.not_to raise_error
```
Comment (ccdredzik) [inline spec/rails_integration_spec.rb:110]: This results in `ParamEvaluator` being called with the following `params`:

```
#<ActionController::Parameters {"filter"=>{"state"=>""}, "controller"=>"fake", "action"=>"nested_array"} permitted: false>
#<ActionController::Parameters {"state"=>""} permitted: false>
```

And succeeds.

File: spec/rails_integration_spec.rb [hunk 2]
```diff
@@ -106,14 +106,12 @@ def prepare_params(params)
...
+    end
 
-      expect { get :nested_array, **prepare_params(params) }.to raise_error do |error|
-        expect(error).to be_a(RailsParam::InvalidParameterError)
-      end
+    it "responds with a 200 when the nested array is processed as nil" do
+      expect { get :nested_array, **prepare_params({ filter: "state" }) }.not_to raise_error
```
Comment (ccdredzik) [inline spec/rails_integration_spec.rb:114]: This results in `ParamEvaluator` being called with the following `params`:

```
#<ActionController::Parameters {"filter"=>"state", "controller"=>"fake", "action"=>"nested_array"} permitted: false>
{"state"=>nil}
```

And fails.

File: spec/rails_integration_spec.rb [hunk 3]
```diff
@@ -106,14 +106,13 @@ def prepare_params(params)
   end
 
   describe "nested_array" do
-    it "responds with a 400 when the nested array is not supplied properly" do
-      params = {
-        'filter' => 'state'
-      }
+    it "responds with a 200 when the nested array is provided as nil" do
+      expect { get :nested_array, **prepare_params({ filter: { state: nil } }) }.not_to raise_error
```
Comment (ccdredzik) [inline spec/rails_integration_spec.rb:110]: this will pass the following to recurrent calls to `ParamEvaluator`:

```
#<ActionController::Parameters {"filter"=>{"state"=>""}, "controller"=>"fake", "action"=>"nested_array"} permitted: false>
#<ActionController::Parameters {"state"=>""} permitted: false>
```

File: spec/rails_integration_spec.rb [hunk 4]
```diff
@@ -106,14 +106,13 @@ def prepare_params(params)
...
+    end
 
-      expect { get :nested_array, **prepare_params(params) }.to raise_error do |error|
-        expect(error).to be_a(RailsParam::InvalidParameterError)
-      end
+    it "responds with a 400 when the nested array is passed as string" do
+      expect { get :nested_array, **prepare_params({ filter: { state: "a string" } }) }
```
Comment (ccdredzik) [inline spec/rails_integration_spec.rb:114]: this will pass the following to recurrent calls to `ParamEvaluator`:

```
#<ActionController::Parameters {"filter"=>{"state"=>"a string"}, "controller"=>"fake", "action"=>"nested_array"} permitted: false>
#<ActionController::Parameters {"state"=>"a string"} permitted: false>
{0=>"a string"}
```

File: spec/rails_integration_spec.rb (hunk 3 above)
Comment (iMacTia) [reply spec/rails_integration_spec.rb]: This looks like Rails getting in the way. Since we're calling a GET endpoint, the parameters are encoded in a query-string format and when decoded the value of state is `''` instead of nil.
We should either test this with a `post` endpoint, or mock `controller.params` like we do in other tests:
```ruby
allow(controller).to receive(:params).and_return({ filter: { state: nil } })
```

File: spec/rails_integration_spec.rb (hunk 4 above)
Comment (iMacTia) [reply spec/rails_integration_spec.rb:115]: My understanding is that this test is expected to return without raising the exception as everything seems valid. Have you added it just for illustration purpose?

File: spec/rails_integration_spec.rb [hunk 5]
```diff
@@ -106,14 +106,12 @@ def prepare_params(params)
   end
 
   describe "nested_array" do
-    it "responds with a 400 when the nested array is not supplied properly" do
```
[the remark this answers is not in the document]
Comment (ifellinaholeonce) [reply spec/rails_integration_spec.rb:109]: What's the value here in removing that test case tho? If the description of the test was inaccurate, let's address that. However, this was an edge case bug that someone encountered and added the specs for, so let's keep that coverage.

File: spec/rails_integration_spec.rb [hunk 6]
```diff
@@ -106,14 +106,14 @@ def prepare_params(params)
...
-      params = {
-        'filter' => 'state'
-      }
+    it "responds with a 400 when the nested array is provided as nil" do
+      expect { get :nested_array, **prepare_params({ filter: { state: nil } }) }
+        .to raise_error(RailsParam::InvalidParameterError)
```
Comment (ifellinaholeonce) [inline spec/rails_integration_spec.rb:111]: This shouldn't be the case, should it? An array should be able to be nil, nested or otherwise. I believe the expectations should be:
- If param is missing/nil and is required, it raises an error
- if param is missing/nil and not required, it stays nil
- if param is missing/nil, not required, and defaults to [], then it's coerced to []

So I believe we should be expecting this to not raise an error. Am I missing something?

File: spec/rails_integration_spec.rb (hunk 4 above)
Comment (ccdredzik) [reply spec/rails_integration_spec.rb:115]: Both of my tests are currently failing. They return 200.

File: spec/rails_integration_spec.rb (hunk 6 above)
Comment (ccdredzik) [reply spec/rails_integration_spec.rb:111]: @ifellinaholeonce sorry for the confusion, this is currently in a bit of a chaos after a few hours of discussion with @iMacTia about how to solve this. The gist is. I implemented those tests first as a 200. They failed. I have applied the change in param_evaluator.rb. They started passing. But the original test started failing. Anyway. I will make them pass now, and I will move to another thread here to discuss that original test.

File: spec/rails_integration_spec.rb (hunk 4 above)
Comment (iMacTia) [reply spec/rails_integration_spec.rb:115]: Ok but this one should return 200, the payload seems valid

Comment (ccdredzik) [reply spec/rails_integration_spec.rb:115]: @iMacTia On that note. In what universe `state: "a string"` coercing to `["a string"]` did sound like a good idea? :-D

File: spec/rails_integration_spec.rb (hunk 6 above)
Comment (iMacTia) [reply spec/rails_integration_spec.rb:111]: This test should fail, but only because of the nested validation on array items. If that was not there, then this would be a valid payload

File: spec/rails_integration_spec.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (ccdredzik) [reply spec/rails_integration_spec.rb:109]: @ifellinaholeonce fair do's. I have assumed that someone just did wrote a rather silly test. I have restored it and it is now failing.

[the remark this answers is not in the document]
Comment (ccdredzik) [reply spec/rails_integration_spec.rb:109]: The `ParamEvaluator` is now receiving the following params:

```
#<ActionController::Parameters {"filter"=>"state", "controller"=>"fake", "action"=>"nested_array"} permitted: false>
{"state"=>nil}
```

[the remark this answers is not in the document]
Comment (ccdredzik) [reply spec/rails_integration_spec.rb:109]: From my understanding `HashParam` will happily coerce `filter: "state"` into `filter: { state: nil }` which will result in the glorious:

```
       expected RailsParam::InvalidParameterError, got #<NoMethodError: undefined method `each_with_index' for nil:NilClass> with backtrace:
         # ./lib/rails_param/param_evaluator.rb:51:in `recurse_on_parameter'
         # ./lib/rails_param/param_evaluator.rb:35:in `param!'
         # ./spec/fixtures/controllers.rb:37:in `block in nested_array'
         # ./lib/rails_param/param_evaluator.rb:66:in `recurse'
         # ./lib/rails_param/param_evaluator.rb:59:in `recurse_on_parameter'
         # ./lib/rails_param/param_evaluator.rb:35:in `param!'
         # ./lib/rails_param/param.rb:3:in `param!'
         # ./spec/fixtures/controllers.rb:36:in `nested_array'
```

[the remark this answers is not in the document]
Comment (iMacTia) [reply spec/rails_integration_spec.rb:109]: `recurse_on_parameter` (or anything before) should raise exception if the parameter is provided and its value is invalid (i.e. `nil`), it shouldn't try to process it.

[the remark this answers is not in the document]
Comment (ccdredzik) [reply spec/rails_integration_spec.rb:109]: @ifellinaholeonce this test is failing, because we've now changed the way we handle nils inside. My initial thought was just to go to recurse_on_parameter and simply add a `return nil if parameter.value.nil?` guard, but that will just make this test case fail, as it will no longer return an error and handle this the same way it handles the below cases. Thoughts?

[the remark this answers is not in the document]
Comment (ccdredzik) [reply spec/rails_integration_spec.rb:109]: I have just applied that change again.

[the remark this answers is not in the document]
Comment (ccdredzik) [reply spec/rails_integration_spec.rb:109]: @iMacTia but should it? If it should, then the next test should equally fail, because I am passing `filter: { state: nil }` explicitly in it, and that's what the current test is processing `filter: "state"` as.

File: spec/rails_integration_spec.rb (hunk 3 above)
Comment (ccdredzik) [reply spec/rails_integration_spec.rb:114]: @iMacTia I have done a test where I changed the `nested_array` to a `post` endpoint and the result is the same:

```
#<ActionController::Parameters {"filter"=>{"state"=>""}, "controller"=>"fake", "action"=>"nested_array"} permitted: false>
#<ActionController::Parameters {"state"=>""} permitted: false>
```

File: spec/rails_integration_spec.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (ccdredzik) [reply spec/rails_integration_spec.rb:109]: @iMacTia on the other hand, if we declare that `filter: "state"` is not valid then on what grounds? On the grounds that `"state"` is not a valid representation of a hash? If so #81 will fix that. If it's on the grounds that `"state"` is a valid representation of a hash, and it should process the hash as `{ state: nil }` and it is an array `state` that shouldn't be `nil` then the whole premise of this PR is reversed, because my whole point is that explicit `nil` should be valid.

[the remark this answers is not in the document]
Comment (iMacTia) [reply spec/rails_integration_spec.rb:109]: Yeah the two are equivalent, the only difference being that the first implicitly sets the array to nil, while on the latter that's explicit. They should both fail on the premise that the controller validates the array content

File: spec/rails_integration_spec.rb (hunk 3 above)
Comment (iMacTia) [reply spec/rails_integration_spec.rb:114]: This is super strange, where is the empty string coming from?!

File: spec/rails_integration_spec.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (ccdredzik) [reply spec/rails_integration_spec.rb:109]: @iMacTia okay, so the grounds for a failure is NOT the fact that the array here is `nil`, but rather the fact that the array here is `nil` when it's definition has the following line:

```
s.param! idx, String, required: true
```

Is that correct? The first two tests that we are talking about here should expect 400 for this:

```
      f.param! :state, Array do |s, idx|
        s.param! idx, String, required: true
      end
```

But should expect 200 for this:

```
      f.param! :state, Array do |s, idx|
        s.param! idx, String, required: true
      end
```

[the remark this answers is not in the document]
Comment (iMacTia) [reply spec/rails_integration_spec.rb:109]: Yep that's how I would expect it to work. I'm not saying that's how it works now, this probably need fixing, but that seems the logical thing to me.

I'm open to other thoughts as well though @ccdredzik @ifellinaholeonce

[the remark this answers is not in the document]
Comment (ifellinaholeonce) [reply spec/rails_integration_spec.rb:109]: Yeah that seems right to me. Thanks for spelling that out @ccdredzik.

[the remark this answers is not in the document]
Comment (ccdredzik) [reply spec/rails_integration_spec.rb:109]: @iMacTia @ifellinaholeonce so I think this now works differently depending on whether `{ array: nil }` was passed explicitly, or was an attribute in a hash coerced from a string that had keys but not values. I will add a bunch of tests first to try to define all of the above cases so we can see what works and what doesn't.

[the remark this answers is not in the document]
Comment (ccdredzik) [reply spec/rails_integration_spec.rb:109]: @iMacTia @ifellinaholeonce for the record: the more I dive into this, the more I think that the param block for array of anything else than a hash was a bad design choice. More sensible representation of the above would be:

```
param! :state, Array, of: String, allow_empty: false
```

I could understand a block if it would be an array of hashes:

```
param! :state, Array, of: Hash, allow_empty: false do |f|
  f.param! :a_param_of_a_hash, String
end
```

[the remark this answers is not in the document]
Comment (iMacTia) [reply spec/rails_integration_spec.rb:109]: Yeah we should support an array of hashes as well, that's pretty much the only meaningful scenario for the block. I also agree the shorthand you show above would make sense, the block seems a bit too much in that case

[the remark this answers is not in the document]
Comment (ccdredzik) [reply spec/rails_integration_spec.rb:109]: @iMacTia and that block is also what makes it problematic to handle nils

[the remark this answers is not in the document]
Comment (ccdredzik) [reply spec/rails_integration_spec.rb:109]: @iMacTia @ifellinaholeonce I have added an awful lot of tests. Can you double check if the assumptions they make are all correct?

[the remark this answers is not in the document]
Comment (ccdredzik) [reply spec/rails_integration_spec.rb:109]: @iMacTia @ifellinaholeonce and a last thing for today: In param_spec.rb on line 472 there is a test case as follows:

```
it 'does not raise exception if array is not required but nested attributes are, and no array is provided' do
```

Which stands in direct contradiction to what we just talked about... I kind of want to give up.

[the remark this answers is not in the document]
Comment (iMacTia) [reply spec/rails_integration_spec.rb:109]: @ifellinaholeonce I've further discussed this with @ccdredzik offline, I agree there have been some confusion after the refactoring around this point. We agreed on dropping this test because, although the premises are good, it does not work well with hash and array coercion, unfortunately.

We all agree that a nil and missing params should be considered in the same way, so it would make sense to skip the recursive coercion when the parameter is provided, but is nil. Exactly as we do with a missing parameter.

This test can potentially be reintroduced later on, but we'll need to write it using a better payload system, as query-strings clearly don't work as expected and make it hard to reproduce the scenario we wanted to test

[the remark this answers is not in the document]
Comment (ccdredzik) [reply spec/rails_integration_spec.rb:109]: I have pushed the implementation to the latest understanding of the contract. Please feel free to review.

**Review by iMacTia — APPROVED** [review]
Quick recap in case something went missing in the long discussion thread:
* This PR follows the rules highlighted [here](https://github.com/nicolasblanco/rails_param/pull/76#issuecomment-928975668)
* This PR introduces a bunch of new tests
* This PR changes 2 existing tests, but we agreed that's the right course of action:
  * [Test 1](https://github.com/nicolasblanco/rails_param/pull/82/files#diff-9bb16d0fbcd8818a4370eb5cbf34e29a78ab90c10d2d217a924ff9b6b4ebe3c3L109): This test was incorrectly setup to begin with. Due to the way hash coercion works, the payload `{ filter: 'state' }` would become `{ filter: { state: nil } }` in the controller params. The test was then failing due to `state` being nil, and not because it was a string.
  * [Test 2](https://github.com/nicolasblanco/rails_param/pull/82/files#diff-9bb16d0fbcd8818a4370eb5cbf34e29a78ab90c10d2d217a924ff9b6b4ebe3c3L133): this was introduced in the refactoring, but breaks the rules highlighted above (`nil` have always been treated as missing param, we should stick with that).

Big thanks @ccdredzik for the extensive research, patience and commitment on getting this rectified and for highlighting different weak points in the current implementation that should be addressed properly on our way towards v2.0 👏 🙏

**Comment by iMacTia** [comment]
@ifellinaholeonce I'm happy to get this merged now, would you like to take one final pass and raise any issue from your perspective before that happens?

**Review by ifellinaholeonce — APPROVED** [review]
This looks great to me! Thanks for all the hard work and back forth on this issue @ccdredzik. It is a tricky subject.

@iMacTia thanks for driving the conversation on this. I'm ready for this to be merged. I think we should make a note of this in the release notes tho. I believe this is a breaking change from the last release, around that array with nested required elements.

**Comment by iMacTia** [comment]
I agree, it should be noted. Though we should also clarify we're effectively reinstating behaviour prior to v1.0 that was unintentionally broken, to justify the minor version bump

**Comment by iMacTia** [comment]
@ifellinaholeonce here is the release draft, please give me a quick feedback (or feel free to change!) when you have a minute: https://github.com/nicolasblanco/rails_param/releases/tag/untagged-b8a8e53920572a12295e
